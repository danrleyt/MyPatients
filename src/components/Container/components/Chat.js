import React from 'react';
import Message from './Message';

const Chat = props => {
  const { messages } = props;
  return (
    <div className="row">
      <div className="messages">
        {messages
          ? messages.map(message => <Message message={message} />)
          : null}
      </div>
      <div className="input">
        <textarea id="textarea1" className="materialize-textarea" />{' '}
        {/* to be implemented */}
      </div>
    </div>
  );
};

export default Chat;
