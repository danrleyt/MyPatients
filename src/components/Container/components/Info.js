import React from 'react';
// import {
//   PhoneNumberUtil as phoneUtil,
//   PhoneNumberFormat as PNF
// } from 'google-libphonenumber';

const Info = props => {
  const { patient } = props;

  const fullName = `${patient.firstName} ${patient.lastName}`;
  const address = `${patient.address.street}, ${patient.address.city}, ${
    patient.address.zip
  }`;
  const today = new Date(); // auxiliar date to calculate age

  const diagnosisMap = {
    // map to show the attribute captalize
    depression: 'Depression',
    eatingDisorder: 'Eating Disorder',
    burnout: 'Burnout'
  };
  const treatmentMap = {
    // map to show the attribute captalize
    outpatient: 'Outpatient',
    aftercare: 'Aftercare'
  };

  return patient ? (
    <div className="row">
      <h5>Name: {fullName}</h5>
      <h5>Address: {address}</h5>
      <h5>Phone number: {patient.phoneNumber}</h5>
      <h5>
        Age: {today.getFullYear() - new Date(patient.dateOfBirth).getFullYear()}{' '}
        years old
      </h5>
      <h5>Insurance Number: {patient.insuranceNumber}</h5>
      <h5>Diagnosis: {diagnosisMap[patient.diagnosis]}</h5>
      <h5>Treatment Type: {treatmentMap[patient.treatmentType]}</h5>
      {patient.emergencyContact ? (
        <div>
          <h5>Emergency Contact </h5>
          <h5>
            {' '}
            Name:{' '}
            {`${patient.emergencyContact.firstName} ${
              patient.emergencyContact.lastName
            }`}
          </h5>
          <h5>Phone Number: {patient.emergencyContact.phoneNumber}</h5>
        </div>
      ) : null}
    </div>
  ) : null;
};

export default Info;
