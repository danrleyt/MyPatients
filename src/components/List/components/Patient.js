import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCircle } from '@fortawesome/free-solid-svg-icons';
import '../../../styles/global.css';
import FemaleAvatar from '../../../res/female-user.png';
import MaleAvatar from '../../../res/male-user.png';

class Patient extends React.Component {
  constructor(props) {
    super(props);
    this.handleImageError = this.handleImageError.bind(this);
    this.passPatient = this.passPatient.bind(this);
  }

  passPatient() {
    const { callback, patient } = this.props;
    callback(patient);
  }

  handleImageError(event) {
    const evt = event;
    const { patient } = this.props;
    evt.preventDefault();
    evt.target.onerror = null;
    evt.target.src = patient.gender === 'female' ? FemaleAvatar : MaleAvatar;
  }

  render() {
    const { patient } = this.props;
    const completeName = `${patient.firstName} ${patient.lastName}`;

    return (
      <div
        className="col s12 card hoverable patient"
        onClick={this.passPatient}
        role="button"
        aria-hidden
      >
        <div className="valign-wrapper ">
          <img
            className="col s3 circle "
            onError={this.handleImageError}
            src={patient.profilePicture}
            alt="profile"
          />
          <div className="col s9 valign-wrapper">
            <h5>{completeName}</h5>
            <FontAwesomeIcon
              icon={faCircle}
              color={patient.online ? 'green' : 'red'}
              pull="right"
            />
          </div>
        </div>
      </div>
    );
  }
}

export default Patient;
