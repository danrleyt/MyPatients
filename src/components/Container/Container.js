import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faComments, faInfoCircle } from '@fortawesome/free-solid-svg-icons';
import Info from './components/Info';
import Chat from './components/Chat';

const borderStyle = {
  border: 'solid',
  height: '100vh',
  overflowY: 'scroll'
};

const icons = {
  margin: 10,
  cursor: 'pointer'
};

class Container extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      showInfo: false // control to check if it should show the info of the patient or the messages
    };
    this.toggleInfo = this.toggleInfo.bind(this);
  }

  toggleInfo(option) {
    this.setState({ showInfo: option });
  }

  render() {
    const { selected } = this.props;
    const { showInfo } = this.state;
    return (
      <div className="col m9 center" style={borderStyle}>
        <div className="header">
          <div className="row">
            <div className="col s5">
              <img
                className="circle right"
                width={50}
                height={50}
                src={selected.profilePicture}
                alt="profile"
              />
            </div>
            <div className="col s3">
              <h5 className="left">{`${selected.firstName} ${
                selected.lastName
              }`}</h5>
            </div>
            <div className="col s4">
              <FontAwesomeIcon
                style={icons}
                className="right"
                icon={faComments}
                size="2x"
                onClick={() => this.toggleInfo(false)}
              />
              <FontAwesomeIcon
                style={icons}
                className="right"
                icon={faInfoCircle}
                size="2x"
                onClick={() => this.toggleInfo(true)}
              />
            </div>
          </div>
        </div>
        {showInfo ? (
          <Info patient={selected} />
        ) : (
          <Chat messages={selected.messages} />
        )}
      </div>
    );
  }
}

export default Container;
