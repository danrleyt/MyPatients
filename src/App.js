import React, { Component } from 'react';
import List from './components/List/List';
import Container from './components/Container/Container';
import Loader from './components/static/Loader';
import Presentation from './components/static/Presentation';
import PatientService from './services/PatientService';

const styleapp = {
  display: 'flex',
  height: '100vh'
};

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      patients: [], // List of patients
      selected: {}, // control which patient was selected
      isLoading: true // control when to show the loader
    };
    this.handleSelectPatient = this.handleSelectPatient.bind(this);
  }

  async componentDidMount() {
    const patientsList = await PatientService.fetchPatients(); // fetch the list of patients
    this.setState({ patients: patientsList, isLoading: false });
  }

  handleSelectPatient(patient) {
    this.setState({ selected: patient });
  }

  render() {
    const { patients, selected, isLoading } = this.state;
    return (
      <div className="row" style={styleapp}>
        {isLoading ? (
          <Loader />
        ) : (
          <List objects={patients} callback={this.handleSelectPatient} />
        )}
        {Object.keys(selected).length ? (
          <Container selected={selected} />
        ) : (
          <Presentation />
        )}
      </div>
    );
  }
}

export default App;
