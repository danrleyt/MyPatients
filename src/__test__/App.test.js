import React from 'react';
import { create } from 'react-test-renderer';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import App from '../App';
import Patient from '../components/List/components/Patient';
import Container from '../components/Container/Container';
import Info from '../components/Container/components/Info';
import Chat from '../components/Container/components/Chat';
import UtilsServices from '../services/UtilsService';
import mockdata from './mockData/example.json';

describe('Basic Flow', () => {
  it('matches the snapshot', () => {
    const component = create(<App />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('shows a list of patients', () => {
    const component = create(<App />);
    const instance = component.getInstance();
    const sorted = UtilsServices.sortList(mockdata);
    const filtered = UtilsServices.defineFilters(sorted);
    instance.setState({ patients: filtered, isLoading: false });
    const { root } = component;
    const listOfPatients = root.findAllByType(Patient);
    expect(listOfPatients.length).toBe(5);
  });

  it('select a patient', async () => {
    const component = create(<App />);
    const instance = component.getInstance();
    const sorted = UtilsServices.sortList(mockdata);
    const filtered = UtilsServices.defineFilters(sorted);
    instance.setState({ patients: filtered, isLoading: false });
    const { root } = component;
    const listOfPatients = root.findAllByType(Patient);
    const { patient } = listOfPatients[0].props;
    listOfPatients[0].children[0].props.onClick();
    const container = root.findAllByType(Container);
    const renderedName = container[0].findAllByType('h5')[0].children[0];
    expect(renderedName).toBe(`${patient.firstName} ${patient.lastName}`);
  });

  it('shows info of the patient', async () => {
    const component = create(<App />);
    const instance = component.getInstance();
    const sorted = UtilsServices.sortList(mockdata);
    const filtered = UtilsServices.defineFilters(sorted);
    instance.setState({ patients: filtered, isLoading: false });
    const { root } = component;
    const listOfPatients = root.findAllByType(Patient);
    listOfPatients[0].children[0].props.onClick();
    const container = root.findAllByType(Container);
    container[0].findAllByType(FontAwesomeIcon)[1].props.onClick();
    expect(container[0].findAllByType(Info).length).toBe(1);
  });

  it('shows messages of the patient', async () => {
    const component = create(<App />);
    const instance = component.getInstance();
    const sorted = UtilsServices.sortList(mockdata);
    const filtered = UtilsServices.defineFilters(sorted);
    instance.setState({ patients: filtered, isLoading: false });
    const { root } = component;
    const listOfPatients = root.findAllByType(Patient);
    listOfPatients[0].children[0].props.onClick();
    const container = root.findAllByType(Container);
    container[0].findAllByType(FontAwesomeIcon)[0].props.onClick();
    expect(container[0].findAllByType(Chat).length).toBe(1);
  });

  it('changes list after filter active', async () => {
    const component = create(<App />);
    const instance = component.getInstance();
    const sorted = UtilsServices.sortList(mockdata);
    const filtered = UtilsServices.defineFilters(sorted);
    instance.setState({ patients: filtered, isLoading: false });
    const { root } = component;
    const filterLinked = root.findAllByType('input');
    const event = {
      preventDefaul() {},
      target: {
        checked: true,
        value: 2
      }
    };
    filterLinked[0].props.onChange(event);
    const listOfPatients = root.findAllByType(Patient);
    expect(listOfPatients.length).toBe(3);
  });
});
