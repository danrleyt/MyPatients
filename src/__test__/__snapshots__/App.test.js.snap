// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component it matches the snapshot 1`] = `
<div
  className="row"
  style={
    Object {
      "display": "flex",
      "height": "100vh",
    }
  }
>
  <div
    className="spinner valign-wrapper"
  >
    <div
      className="rect1"
    />
    <div
      className="rect2"
    />
    <div
      className="rect3"
    />
    <div
      className="rect4"
    />
    <div
      className="rect5"
    />
  </div>
  <div
    className="col m9 center"
  >
    <div>
      <h2>
        Welcome to MyPatients App
      </h2>
      <br />
      <h3>
        Select a Patient on the menu at the left
      </h3>
    </div>
  </div>
</div>
`;

exports[`App Component matches the snapshot 1`] = `
<div
  className="row"
  style={
    Object {
      "display": "flex",
      "height": "100vh",
    }
  }
>
  <div
    className="spinner valign-wrapper"
  >
    <div
      className="rect1"
    />
    <div
      className="rect2"
    />
    <div
      className="rect3"
    />
    <div
      className="rect4"
    />
    <div
      className="rect5"
    />
  </div>
  <div
    className="col m9 center"
  >
    <div>
      <h2>
        Welcome to MyPatients App
      </h2>
      <br />
      <h3>
        Select a Patient on the menu at the left
      </h3>
    </div>
  </div>
</div>
`;

exports[`Basic Flow matches the snapshot 1`] = `
<div
  className="row"
  style={
    Object {
      "display": "flex",
      "height": "100vh",
    }
  }
>
  <div
    className="spinner valign-wrapper"
  >
    <div
      className="rect1"
    />
    <div
      className="rect2"
    />
    <div
      className="rect3"
    />
    <div
      className="rect4"
    />
    <div
      className="rect5"
    />
  </div>
  <div
    className="col m9 center"
  >
    <div>
      <h2>
        Welcome to MyPatients App
      </h2>
      <br />
      <h3>
        Select a Patient on the menu at the left
      </h3>
    </div>
  </div>
</div>
`;
